util/election/runner.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+// Config is a set of parameters for a continuous master election process.
+type Config struct {
+	// PreElectionPause is the maximum interval to wait before starting election.
+	PreElectionPause time.Duration
+	// MasterCheckInterval is the interval between checks that the instance still
+	// holds mastership.
+	MasterCheckInterval time.Duration
+	// MasterTTL is the time after which the instance will resign if encountered
+	// no successful IsMaster calls.
+	MasterTTL time.Duration
+}
+
+// Run contains mastership status of a single instance during a single run of
+// being master.
+type Run struct {
+	// Start is the time when the instance became the master as measured by the
+	// corresponding Runner's TimeSource.
+	Start time.Time
+	// Ctx is the context which is active while the instance believes to be the
+	// master. As soon as Ctx is canceled, the instance attempts to resign, and
+	// stops monitoring its status.
+	Ctx context.Context
+	// Cancel cancels Ctx, or does nothing if that is already canceled.
+	// Effectively, it instructs the instance to resign if not yet.
+	Cancel context.CancelFunc
+	// Done closes when mastership status monitoring goroutine for this run
+	// terminates, strictly after Ctx.Done() closes. After calling Cancel, a
+	// closed Done indicates completion of resigning (unless it failed, or Ctx
+	// parent context was canceled earlier).
+	Done <-chan struct{}
```

Comment:
Do all these fields need to be exported?  (Or even here: `Start` and `Ctx` might not be needed.)

Reply:
`Start` is convenient for the client code to apply their resigning policies based on mastership time. Altough the time could be measured by the client entirely.

I intended `Ctx` to be used by "while-master" jobs, please see the example snippet above `BeMaster`. In Trillian, for instance, each log signing iteration could be run with this `Ctx` so as to stop abruptly when mastership is lost.